linkerd/http/retry/src/lib.rs

```diff
@@ -81,31 +83,45 @@ struct Metrics {
 
 // === impl NewHttpRetry ===
 
-impl<P, L: Clone, N> NewHttpRetry<P, L, N> {
-    pub fn layer(metrics: MetricFamilies<L>) -> impl layer::Layer<N, Service = Self> + Clone {
+impl<P, X, L: Clone, F: Clone, N> NewHttpRetry<P, L, F, X, N> {
+    pub fn layer_via_mk(
+        mk_extract: F,
+        metrics: MetricFamilies<L>,
+    ) -> impl tower::layer::Layer<N, Service = Self> + Clone {
         layer::mk(move |inner| Self {
             inner,
+            mk_extract: mk_extract.clone(),
             metrics: metrics.clone(),
             _marker: PhantomData,
         })
     }
 }
 
-impl<T, P, L, N> NewService<T> for NewHttpRetry<P, L, N>
+impl<T, X, P, L, F, N> NewService<T> for NewHttpRetry<P, L, F, X, N>
 where
-    T: Param<L>,
     P: Policy,
     L: Clone + std::fmt::Debug + Hash + Eq + Send + Sync + prom::encoding::EncodeLabelSet + 'static,
+    F: Fn(&T) -> X + Clone,
```

Comment:
Eh, this gives me a bit of pause... we're adding an unnamable type parameter here. I think this is basically an ExtractParam?

Reply:
Note https://github.com/linkerd/linkerd2-proxy/blob/89ba5b5d8b39f54de8a8c6e1e8cf552e9c9a5527/linkerd/stack/src/lib.rs#L115-L122

Reply:
To keep things ~consistentish, I'd be inclined to use `X: ExtractParam<T, ReqX>` in place of `F`--since X tends to be used for a when a NewService needs an ExtractParam, and then have a secondary `ReqX` (or something) for the second level of extraction.

Reply:
750bd6e renames `F` and `X` to `X` and `ReqX`, respectively. 323dedb changes the trait bounds to use the blanket implementation linked above. good call!